Correct humidity controls for dehumidification - Contoller:WaterCoil and SetpointManager:SingleZone:Humidity:Maximum

rraustad opened the pull request:
Addresses #4994. Humidity control (TemperatureAndHumidityRatio) could cause Controller:WaterCoil to request colder outlet air temperatures than necessary when the coil outlet humidity ratio was already lower than the humidity ratio set point.
Also improved SetpointManager:SingleZone:Humidity:Maximum to always place a setpoint regardless of the sign of Zone Predicted Moisture Load to Dehumidifying Setpoint Moisture Transfer Rate.  Previously, if this load was positive, the setpoint would be set to zero.  This could cause poor RH control at certain times for cooling  coils, desiccant dehumidifiers and other equipment.

rraustad committed:
Correct TemperatureAndHumidityRatio control for water coils

rraustad commented on src/EnergyPlus/HVACControllers.cc:
```diff
@@ -1130,49 +1145,49 @@ namespace HVACControllers {
-					ControllerProps( ControlNum ).SetPointValue = Node( SensedNode ).TempSetPoint;
-					// Finally indicate thate the setpoint has been computed
-					ControllerProps( ControlNum ).IsSetPointDefinedFlag = true;
-				}
```
Problem here was that Tsp was calculated only once each time step, and at full water flow rate. Opened this up to calculate Tsp each iteration. This takes longer to converge, but gets closer to the correct answer.

rraustad commented on src/EnergyPlus/HVACControllers.cc:
```diff
@@ -1130,49 +1145,49 @@ namespace HVACControllers {
+				ControllerProps( ControlNum ).SaveTempSetpointValue = Node( SensedNode ).TempSetPoint;
+				ControllerProps( ControlNum ).SaveTempSetpoint = false;
+			}
+			// Setpoint temp calculated each HVAC step to identify approach temp and whether or not humrat control is necessary
```
Save the sensible only Tsp for use during subsequent iterations.

rraustad commented on src/EnergyPlus/HVACControllers.cc:
```diff
@@ -1130,49 +1145,49 @@ namespace HVACControllers {
+				//       once when the air loop has been evaluated with the max actuated value.
+				//       See routine CalcSimpleController() for the sequence of operations.
+				ControllerProps( ControlNum ).SetPointValue = min( ControllerProps( ControlNum ).SaveTempSetpointValue, HumidityControlTempSetPoint ); // Pure temperature setpoint | Temperature setpoint to achieve the humidity ratio setpoint
+				// Overwrite the "pure" temperature setpoint with the actual setpoint that takes into
```
Compare the humrat Tsp with the original sensible only Tsp

rraustad commented on src/EnergyPlus/HVACControllers.cc:
```diff
@@ -1130,49 +1145,49 @@ namespace HVACControllers {
+				//       controller iteration. For this reason we perform the setpoint calculation only
+				//       once when the air loop has been evaluated with the max actuated value.
+				//       See routine CalcSimpleController() for the sequence of operations.
+				ControllerProps( ControlNum ).SetPointValue = min( ControllerProps( ControlNum ).SaveTempSetpointValue, HumidityControlTempSetPoint ); // Pure temperature setpoint | Temperature setpoint to achieve the humidity ratio setpoint
```
Compare the humrat Tsp with the original sensible only Tsp

Myoldmopar replied:
@rraustad are you working while in line waiting for a great black Friday deal??

rraustad replied:
Home and full of turkey. This one bugged me so I had to fix it.

rraustad commented on tst/EnergyPlus/unit/HVACControllers.unit.cc:
```diff
@@ -91,13 +91,74 @@ namespace EnergyPlus {
-		ControllerLists.deallocate();
-		AllSetPtMgr.deallocate();
-		SchSetPtMgr.deallocate();
-		WaterCoil.deallocate();
```
Most were clear_state, added clear_state for ControllerProps.

rraustad commented:
First file with diffs is 5ZoneElectricBaseboard...  
    Controller:WaterCoil,
     Central Cooling Coil Contoller 1,  !- Name
     TemperatureAndHumidityRatio,  !- Control Variable

All other files with diff's also use TemperatureAndHumidityRatio as the control variable.

rraustad commented:
Comparison of 5ZoneElectricBaseboard using V8.4 and revised develop code.

Uses:
    SetpointManager:SingleZone:Humidity:Maximum,
      Space2-1 Humidistat Manager,  !- Name
      ,                        !- Control Variable
      ,                        !- Schedule Name
      Main Cooling Coil 1 Outlet Node,  !- Setpoint Node or NodeList Name
      SPACE2-1 Node;           !- Control Zone Air Node Name

Node Humidity Ratio Maximum is 0 unless humidity control is needed.
RH SP = 50%. New results show nearly identical control.

Not sure why humrat max (yellow) is not > 0 when RH is over 60% in middle of second run period (about 3/4 of way through data)

![5zonebb_8 4](https://cloud.githubusercontent.com/assets/5778156/11446697/8c02969c-9508-11e5-9a3a-1ae73cf479a1.png)

![5zonebb_new](https://cloud.githubusercontent.com/assets/5778156/11445513/99a220f0-94fa-11e5-9e16-cdb9e1eec68c.png)

rraustad commented:
RH > 50 cannot be controlled if system airflow = 0.

![5zonebb_airflow](https://cloud.githubusercontent.com/assets/5778156/11446047/0da49e42-9500-11e5-9f16-21a49026b845.png)

rraustad committed:
Reviewed IORef and revised unit test

mjwitte commented on src/EnergyPlus/HVACControllers.cc:
```diff
@@ -1130,49 +1145,49 @@ namespace HVACControllers {
+			//       - Node(SensedNode)%HumRatMax
+			//       - Node(SensedNode)%Temp
+			//       - Node(SensedNode)%HumRat
+			if ( Node( SensedNode ).HumRatMax > 0 ) {
```
@rraustad Seems this should also check if the entering HumRat is > Node(SensedNode).HumRatMax.  No need to figure out an adjusted setpoint if the entering air is already drier than the target HumRatMax.

rraustad replied:
I did consider testing that, and regardless the ControllerProps.SetPointValue will have to be set. It seemed to me that the min function for ControllerProps.SetPointValue at line 1179 takes care of HumRat>HumRatMax. So my thoughts were, do I add another test in the if at line 1164 and also add an ELSE at the end to set the sensible only SP, or do I always pass through this IF with the expectation that the min function will always result in the correct answer. You can see which thought prevailed.

mjwitte replied:
If I'm following correctly, you already have the ELSE here, so it seems if the sensed node .HumRat is < .HumRatMax you can skip two psych calls.

rraustad replied:
If I were to change this, it would have to look like this since some SPMs set HumRatMax = 0 when no dehumidification is required. The defect file used a scheduled SPM so I could see where some clock cycle could be saved. Are you OK with this?

```
if ( Node( SensedNode ).HumRatMax > 0 && Node( SensedNode ).HumRat > Node( SensedNode ).HumRatMax ) {

} else {
    // Pure temperature setpoint control strategy
    ControllerProps( ControlNum ).SetPointValue = ControllerProps( ControlNum ).SaveTempSetpointValue;
    // Finally indicate thate the setpoint has been computed
    ControllerProps( ControlNum ).IsSetPointDefinedFlag = true;
}
```

mjwitte replied:
Yes, I expected to keep the >0 condition and check both.  I would add more parens just for clarity.

rraustad replied:
I made that change and tested the defect file. Results look promising. If you do run the defect file, lower the HumRat SP from 0.0075 to at least 0.006 or you won't see any humidity control. I got rid of all variables except for 

  Output:Variable,VAV 1 Cooling Coil Outlet,System Node Temperature,hourly;
  Output:Variable,VAV 1 Cooling Coil Outlet,System Node Setpoint Temperature,hourly;
  Output:Variable,VAV 1 Cooling Coil Outlet,System Node Setpoint Maximum Humidity Ratio,hourly;
  Output:Variable,VAV 1 Cooling Coil Outlet,System Node Humidity Ratio,hourly;

  Output:Variable,VAV 2 Cooling Coil Outlet,System Node Temperature,hourly;
  Output:Variable,VAV 2 Cooling Coil Outlet,System Node Setpoint Temperature,hourly;
  Output:Variable,VAV 2 Cooling Coil Outlet,System Node Setpoint Maximum Humidity Ratio,hourly;
  Output:Variable,VAV 2 Cooling Coil Outlet,System Node Humidity Ratio,hourly;

  Output:Variable,VAV 3 Cooling Coil Outlet,System Node Temperature,hourly;
  Output:Variable,VAV 3 Cooling Coil Outlet,System Node Setpoint Temperature,hourly;
  Output:Variable,VAV 3 Cooling Coil Outlet,System Node Setpoint Maximum Humidity Ratio,hourly;
  Output:Variable,VAV 3 Cooling Coil Outlet,System Node Humidity Ratio,hourly;

rraustad committed:
Slight mod to IF test for when to calculate new SP based on humidity

mjwitte commented:
@rraustad I like the stronger IF - just seems cleaner.  Will review CI results after this reports back.

mjwitte committed:
Merge remote-tracking branch 'remotes/origin/develop' into #4994-ControllerWaterCoil-TemperatureAndHumidityRatio-control-active-when-it-shouldn't-be

mjwitte commented:
Merged in develop (yesterday) and tested a couple of annual runs, 5ZoneAirCooled-Annual and 5ZoneWaterCooled_HighRHControl-Annual, looking for any change in speed (due to moving some of the controller code to every itereation) and results differences.
5ZoneAirCooled - no diffs in results, and similar runtimes.
5ZoneWaterCooled_HighRHControl - got some new errors that are not present with current develop (as of yesterday)

```
   ** Warning ** SolveAirLoopControllers: Maximum iterations (50) exceeded for VAV SYS 1, at Chicago Ohare Intl Ap IL USA TMY3 WMO#=725300, 02/23 19:15 - 19:30


   *************  ** Warning ** SolveAirLoopControllers: Exceeding Maximum iterations for VAV SYS 1 during Chicago Ohare Intl Ap IL USA TMY3 WMO#=725300 continues
   *************  **   ~~~   **   This error occurred 61 total times;
   *************  **   ~~~   **   during Warmup 0 times;
   *************  **   ~~~   **   during Sizing 0 times.
```

5ZoneWaterCooled_HighRHControl-Annual has some small <1% diffs in annual results, and small diffs in hourly results.

Will try a build with just the IF change and not moving code to fire every iteration.

mjwitte committed:
Revert controller Tsp calc back to once per time step

mjwitte committed:
Merge remote-tracking branch 'remotes/origin/develop' into #4994-ControllerWaterCoil-TemperatureAndHumidityRatio-control-active-when-it-shouldn't-be

mjwitte committed:
Fix indentation in HVACControllers.cc

rraustad commented:
The good news is...that error is not a failure of the controller ... it's a failure of the air loop convergence (OK, or a response of the controller to the action of the air loop). I've had a suspicion for some time that there might be an issue with how supply/demand side node conditions are saved, compared, and pushed across the connection. Have not tried to track that down since it is rare.

rraustad commented:
I am just so uncomfortable with this:

"// Setpoint can only be computed when the sensed node temperature is evaluated at the max"

I get it that the controller is picky, but that's just wrong. I want to get rid of ControlCompOutput...

All this info is good to move towards a solution. Thanks.

mjwitte commented:
@rraustad I think these comments a few lines further down may hold the explanation.

```
                        // NOTE: The next line introduces a potential discontinuity into the residual function 
1160 1174                       //       which could prevent the root finder from finding the root it if were done at each 
1161 1175                       //       controller iteration. For this reason we perform the setpoint calculation only 
1162 1176                       //       once when the air loop has been evaluated with the max actuated value. 
1163 1177                       //       See routine CalcSimpleController() for the sequence of operations. 
```

If you follow the blame in EnergyPlusArchive, you'll probably find this was added to fix a problem.  @wfbuhl may remember?

mjwitte commented:
Once the CI reports come back, if there are no problems, I'm inclined to merge this as-is.

rraustad commented:
It looks like these diff's are due to warmup convergence diffs (see eio). I would expect that with a change in the humrat SP temp.

Myoldmopar commented:
Yeah, this looks good to go in.  I tried to follow the blame back on the archive, but there was a big commit that incidentally touched all the files in a big way and basically forced a re-write.  So it's more difficult to trace back.  I would be willing to dig further if we need to open a follow-up ticket for that.

mjwitte commented:
@rraustad @Myoldmopar Not all of the diffs are just warmup convergence, VSHeatPumpWaterHeater for one shows a significant reduction in cooling and heating energy with this fix.  And adding more output reveals that after this fix, the zone RH is no longer being controlled as well as before.  Back to the drawing board.  Will add the VSHeatPumpWaterHeater with additional output to the defect folder.

mjwitte commented:
@rraustad I'm making a build off of your original fix (with Tsp updates every iteration) to see if I get the same result.

mjwitte commented:
@rraustad Going back to your original code puts VSHeatPumpWaterHeater back where it should be - just minor diffs. Regarding the original code for this section, is was introduced in 2006 as part of an [Air loop simulation speedup](https://github.com/NREL/EnergyPlusArchive/commit/1a2af8e7f81dd63436c8fc8968e1fd65db1d97ed).  There are still comments in the current code about [these changes](https://github.com/NREL/EnergyPlus/blob/726bfb957952578c3923f9b97ed342280e4c2e51/src/EnergyPlus/HVACControllers.cc#L42-L76) and it's quite specific about wanting to adjust these setpoints just once per timestep.  And BTW, @Myoldmopar , these comments also mention the secret controller tracking output discussed in #5350 

So, at this point, I'm uncertain if the original fix is the best path forward or some other change.  Clearly the current state of this branch does not work.

rraustad commented:
Push back the original fix so I can look at it again. I think I just need to initialize the controller each iteration, and only for controllers using iTemperatureAndHumidityRatio so we don't loose speed improvements for most input files.

mjwitte committed:
Restore controller Tsp calc every timestep (original RR fix)

rraustad commented:
Thanks

mjwitte commented:
OK - hope I got it all back how it was.

rraustad commented:
I added a change in SP to an extreme, both in the positive and negative direction. Doing this after the initial SP was set. No root controller errors. So why are we basing our decisions on code comments when those comments might not take into account things their crystal ball didn't predict at the time?

```
if( ControllerProps( ControlNum ).IsSetPointDefinedFlag ) {
    ControllerProps( ControlNum ).SetPointValue = -100.0;
}
```

Lets get back to why the original fix wasn't considered. Was is just the code comments?

rraustad commented:
Spoke too soon, now I have something to work with.

```
** Severe  ** FindRootSimpleController: Controller error for controller = "CENTRAL COOLING COIL CONTOLLER 1"
**   ~~~   **  Environment=MIAMI INTL AP ANN HTG 99.6% CONDNS DB, at Simulation time=01/21 09:45 - 10:00
**   ~~~   **   Controller function is inconsistent with user specified controller action = Reverse action
**   ~~~   **   Actuator will be set to maximum action
**   ~~~   ** Controller control type=Temperature and humidity ratio
**   ~~~   ** Controller temperature setpoint = -100.00 [C]
```

mjwitte commented:
@rraustad Original reason for reverting back to once per timestep was this:

5ZoneWaterCooled_HighRHControl - got some new errors that are not present with current develop (as of yesterday)

```
   ** Warning ** SolveAirLoopControllers: Maximum iterations (50) exceeded for VAV SYS 1, at Chicago Ohare Intl Ap IL USA TMY3 WMO#=725300, 02/23 19:15 - 19:30


   *************  ** Warning ** SolveAirLoopControllers: Exceeding Maximum iterations for VAV SYS 1 during Chicago Ohare Intl Ap IL USA TMY3 WMO#=725300 continues
   *************  **   ~~~   **   This error occurred 61 total times;
   *************  **   ~~~   **   during Warmup 0 times;
   *************  **   ~~~   **   during Sizing 0 times.
```

Plus a concern about touching more than necessary and possible slowdown.

mjwitte commented:
That's as an annual run.

rraustad commented:
This looks like pretty good RH control. Just need to figure out why a warning is being reported.

![image](https://cloud.githubusercontent.com/assets/5778156/11734066/b2486e3e-9f81-11e5-95bb-7d433b187233.png)

rraustad commented:
Some interesting developments.

1) the max humrat SP manager does not set the humrat max SP if the moisture load > 0. This is wrong since if the coil overshoots slightly, which it can since it converges using a tolerance, the coil operates at the sensible TSP. Changing to always calculate the humrat max SP yields this:

![5zonewatercooled_highrhcontrol](https://cloud.githubusercontent.com/assets/5778156/11759718/99ebb3cc-a04f-11e5-8224-fe80ec80b643.png)

2) I see some SolveAirLoopControllers: Maximum iterations warnings. I think these occur when the coil outlet air humrat doesn't change due to a dry coil condition. In the water coil code, the model only calculates a new outlet air hum rat if the difference between the air DP and calculated condensation point is > 0.1. I changed this to 0.01 and the warnings reduced from 42 to 13. I assume the 13 remaining are a result of a dry coil.

if ( ( TempAirDewPoint - TempCondensation ) > 0.1 ) {

3) There doesn't seem to be a problem with changing the SP during subsequent iterations. The SP just can't be set lower than the coil inlet water temperature.

I'm stuck now with SolveAirLoopControlers: Max Iteration warnings that I think are due to a dry coil (winter months, 1-3 and 11-12). Not sure what I can do about that. 

Update: I'm using Miami weather on this Chicago example file.

Update 2: What looks like is happening is that since the coil outlet air humrat isn't changing, yet the T outlet is, the approach temp calculated in the TemperatureAndHumidityRatio code is slightly changing the TSP each iteration... and the controller goes to max iterations.

rraustad committed:
Revert to once SP per time step for water coils

rraustad commented:
I give up. The HVACController is just too picky.
Reverted to calculate SP only once per time step.
Still calculating HumRatMax SP all the time, even when MoistureLoad > 0.
Reverted water coil model tolerance to 0.1

Example file, 5ZoneWaterCooled_HighRHControl, with Miami weather shows better control, mostly.
![5zonewatercooled_highrhcontrol2](https://cloud.githubusercontent.com/assets/5778156/11763916/94a28f84-a0eb-11e5-9951-31b5628477cd.png)

rraustad commented:
![image](https://cloud.githubusercontent.com/assets/5778156/11764297/2a95af06-a0f8-11e5-8dd7-e7f076ad2908.png)

mjwitte committed:
Merge remote-tracking branch 'remotes/origin/develop' into #4994-ControllerWaterCoil-TemperatureAndHumidityRatio-control-active-when-it-shouldn't-be

mjwitte commented on src/EnergyPlus/HVACControllers.cc:
```diff
@@ -1145,30 +1159,30 @@ namespace HVACControllers {
 					// Setpoint can only be computed when the sensed node temperature is evaluated at the max
 					// actuated value for the dual humidity ratio / temperature strategy.
 					// See routine CalcSimpleController() for the sequence of operations.
-					if ( ControllerProps( ControlNum ).NextActuatedValue == RootFinders( ControlNum ).MaxPoint.X ) {
```
@rraustad With this if removed, are the above comments still relevant?  ". . . is evaluated at the max actuated value"

rraustad replied:
This comment should be revised to describe revised code. I'll do that after all comments are received.

mjwitte commented:
@rraustad I've tested this with the files I was looking at before (annuals and the defect files) and they all look Ok with this fix.  I'm going to merge in develop one more time to see what other diffs are showing.  I am mildly uncomfortable with changing the setpoint manager to always set the SupplyAirHumRat.  Did you search the code for any other components that use that setpoint and need it to be zero at times?

rraustad commented:
I didn't look for code that expected a 0 to mean no dehumidification. I hoped that the test suite would show which files had issues and point to areas of the code that needed to be investigated. Line 5175 shows that a supply air humidity ratio is calculated based on the zone node hum rat and the predicted moisture load. If the moisture load is positive it simply means the supply air hum rat should be higher than the zone node hum rat, and that supply air hum rat will be needed to maintain the humidistat SP. However, your hesitance is justified.

rraustad commented:
The only one I see so far.

WindowAC:

```
// If the QZnReq <= FullOutput and a HXAssisted coil is used, check the node setpoint for a maximum humidity ratio set piont
// HumRatMax will either equal -999 if no setpoint exists or could be 0 if no moisture load is present
if ( QZnReq <= FullOutput && WindAC( WindACNum ).DXCoilType_Num == CoilDX_CoolingHXAssisted && Node( WindAC( WindACNum ).CoilOutletNodeNum ).HumRatMax <= 0.0 ) {
```

Should be changed to use Node.HumRatMax >= Node.HumRat

rraustad commented:
The min hum rat SPM also always sets the hum rat min SP if there is flow.
DefineSZMinHumSetPointManager::calculate()

mjwitte commented:
@rraustad Interesting, so the min humrat SP manager already sets it regardless of the moisture load.  That makes me more comfortable with this.

mjwitte commented:
CI results show diffs in test files with RH control on water coils as expected.  There are also diffs in desiccant dehumidifier test files, such as DesiccantCVRhZoneRHCtrl.  Looking at this file in more details shows that setting the max humrat SP every time step results in better control.  Before this fix is on the left.  Note how the zone RH is not maintained from 7-8 because the dehumidifier is cycling off every few timesteps.  With the fix, it now runs every timestep to maintain the RH setpoint.

![image](https://cloud.githubusercontent.com/assets/4991075/11812826/c933424c-a303-11e5-9a64-8a7cc1cfa84e.png)

So, the setpointmanager change is good.  

Only question left at this point is whether we need to change the WindowAC controls.  No, it's ok as-is, because windowAC is working directly off the zone humidistat load, it doesn't use setpoints.

@rraustad My only remaining item is to update those comments about max flow.
@Myoldmopar Any further comments?

rraustad commented:
Comparison of WindACAuto with an added flat plate HX.
RH SP is 47.5, you can see the HX turning on and off.
Uses the suggestion above where HumRatMax <= 0 is changed to Node.HumRatMax >= Node.HumRat
Zoomed in to see zone RH and delta T across one side of HX.
This looks good, I'll push this up now.

![windacwhx](https://cloud.githubusercontent.com/assets/5778156/11813954/3a4af9bc-a312-11e5-82fb-90e437fdb881.png)

rraustad committed:
Correct WindowAC RH control based on recent changes to SPM HumRatMax

rraustad commented:
FYI, the CoilSystem HX needs a HumRatMax SP to turn the HX on and off based on coil outlet node HumRat, at least in the WindowAC model. Otherwise, the HX is always on as shown below (SPM removed and there is no warning).

![image](https://cloud.githubusercontent.com/assets/5778156/11814387/744f44ae-a314-11e5-87c8-2396b609a3ce.png)

rraustad commented:
OK, wait a minute, the last figure above is in V8.4. I ran this in develop and I don't get the same answer. I'll investigate. I think I need to add another check for if SP == -9999

rraustad committed:
Reverted WindowAC changes since they were already adequate for RH control

rraustad commented:
OK, you were right that the WindowAC model didn't need to be changed. It does look at outlet air HumRatMax, but the logic still worked. My change didn't work if there was no outlet node HR SPM. It still amazes my how much testing is needed for a simple code change. Now the CoilSystem version works the same as V8.4, with and without the HumRatMax SPM.

mjwitte commented:
@Myoldmopar I think this is ready to go in, either now or after bug release.
